File: apps/frontend/src/api/modules/analytics.ts
import { env } from '@/shared/config/env';

import { apiRequest } from '@/api/http-client';

const AUTH_STORAGE_KEY = 'atmos.auth.v1';

function getStoredAccessToken(): string | null {
  if (typeof window === 'undefined') {
    return null;
  }

  const raw = window.localStorage.getItem(AUTH_STORAGE_KEY);
  if (!raw) {
    return null;
  }

  try {
    const parsed = JSON.parse(raw) as { accessToken?: string };
    return parsed.accessToken ?? null;
  } catch {
    return null;
  }
}

export interface AnalyticsSourceOption {
  id: number;
  name: string;
  source_type: string;
  etl_run_id: string;
  downloaded_at: string | null;
  row_count: number;
  variable_codes: string[];
  period_start: string | null;
  period_end: string | null;
  downloaded_by: string | null;
}

interface AnalyticsStationOption {
  code: string;
  name: string;
  latitude: number | null;
  longitude: number | null;
  region: string | null;
}

interface AnalyticsVariableOption {
  code: string;
  name: string;
}

export interface AnalyticsFilterOptionsResponse {
  sources: AnalyticsSourceOption[];
  stations: AnalyticsStationOption[];
  variables: AnalyticsVariableOption[];
  min_observed_at: string | null;
  max_observed_at: string | null;
}

export interface AnalyticsQueryRequest {
  source_file_ids?: number[];
  station_codes?: string[];
  variable_codes?: string[];
  date_from?: string;
  date_to?: string;
  view_from?: string;
  view_to?: string;
  limit?: number;
}

export interface AnalyticsDataRow {
  observed_at: string;
  station_code: string;
  station_name: string;
  variable_code: string;
  variable_name: string;
  value: number;
  unit: string | null;
  source_file_id: number;
  source_file_name: string;
  source_type: string;
}

export interface AnalyticsQueryResponse {
  rows: AnalyticsDataRow[];
  row_count: number;
  truncated: boolean;
}

export function getAnalyticsFilters(): Promise<AnalyticsFilterOptionsResponse> {
  return apiRequest<AnalyticsFilterOptionsResponse>('/api/v1/analytics/filters');
}

export function runAnalyticsQuery(payload: AnalyticsQueryRequest): Promise<AnalyticsQueryResponse> {
  return apiRequest<AnalyticsQueryResponse>('/api/v1/analytics/query', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export async function exportAnalyticsQuery(payload: AnalyticsQueryRequest): Promise<{ blob: Blob; filename: string }> {
  const accessToken = getStoredAccessToken();
  const response = await fetch(`${env.apiBaseUrl}/api/v1/analytics/export`, {
    method: 'POST',
    body: JSON.stringify(payload),
    headers: {
      'Content-Type': 'application/json',
      ...(accessToken ? { Authorization: `Bearer ${accessToken}` } : {}),
    },
  });

  if (!response.ok) {
    let detail = `Export failed: ${response.status}`;
    try {
      const errorPayload = (await response.json()) as { detail?: string };
      if (errorPayload.detail) {
        detail = errorPayload.detail;
      }
    } catch {
      // Keep fallback detail
    }
    throw new Error(detail);
  }

  const disposition = response.headers.get('Content-Disposition') ?? '';
  const encodedMatch = /filename\*=UTF-8''([^;]+)/i.exec(disposition);
  const quotedMatch = /filename="?([^"]+)"?/i.exec(disposition);
  const filename = encodedMatch?.[1]
    ? decodeURIComponent(encodedMatch[1])
    : quotedMatch?.[1] ?? 'remmaq-records.csv';

  return { blob: await response.blob(), filename };
}

File: apps/backend/app/schemas/analytics.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, Field


class AnalyticsSourceOption(BaseModel):
    id: int
    name: str
    source_type: str
    etl_run_id: str
    downloaded_at: datetime | None
    row_count: int


class AnalyticsStationOption(BaseModel):
    code: str
    name: str


class AnalyticsVariableOption(BaseModel):
    code: str
    name: str


class AnalyticsFilterOptionsResponse(BaseModel):
    sources: list[AnalyticsSourceOption]
    stations: list[AnalyticsStationOption]
    variables: list[AnalyticsVariableOption]
    min_observed_at: datetime | None
    max_observed_at: datetime | None


class AnalyticsQueryRequest(BaseModel):
    source_file_ids: list[int] = Field(default_factory=list)
    station_codes: list[str] = Field(default_factory=list)
    variable_codes: list[str] = Field(default_factory=list)
    date_from: date | None = None
    date_to: date | None = None
    limit: int = Field(default=5000, ge=100, le=20000)


class AnalyticsDataRowResponse(BaseModel):
    observed_at: datetime
    station_code: str
    station_name: str
    variable_code: str
    variable_name: str
    value: float
    unit: str | None
    source_file_id: int
    source_file_name: str
    source_type: str


class AnalyticsQueryResponse(BaseModel):
    rows: list[AnalyticsDataRowResponse]
    row_count: int
    truncated: bool


class SqlPreviewRequest(BaseModel):
    sql: str
    limit: int = Field(default=120, ge=1, le=500)


class SqlPreviewResponse(BaseModel):
    columns: list[str]
    rows: list[dict[str, Any]]
    row_count: int
    truncated: bool
